components/motor/gpio/loop.go

```diff
@@ -60,27 +75,259 @@ func (m *EncodedMotor) setupControlLoop() error {
+	cm.mu.RLock()
+	defer cm.mu.RUnlock()
+	return (ticks + cm.offsetInTicks) / cm.ticksPerRotation, nil
+}
+
+// ResetZeroPosition sets the current position (+/- offset) to be the new zero (home) position.
+func (cm *controlledMotor) ResetZeroPosition(ctx context.Context, offset float64, extra map[string]interface{}) error {
+	// Stop the motor if resetting position
+	if err := cm.Stop(ctx, extra); err != nil {
+		return err
+	}
+	if err := cm.enc.ResetPosition(ctx, extra); err != nil {
+		return err
+	}
+
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	cm.offsetInTicks = -1 * offset * cm.ticksPerRotation
+	return nil
+}
+
+// GoTo instructs the motor to go to a specific position (provided in revolutions from home/zero),
+// at a specific speed. Regardless of the directionality of the RPM this function will move the motor
+// towards the specified target/position
+// This will block until the position has been reached.
+func (cm *controlledMotor) GoTo(ctx context.Context, rpm, targetPosition float64, extra map[string]interface{}) error {
+	// no op manager added, we're relying on GoFor's oepration manager in this driver
+	pos, err := cm.Position(ctx, extra)
+	if err != nil {
+		return err
+	}
+	currRotations := pos / cm.ticksPerRotation
+	rotations := targetPosition - currRotations
+	// if you call GoFor with 0 revolutions, the motor will spin forever. If we are at the target,
+	// we must avoid this by not calling GoFor.
+	if rdkutils.Float64AlmostEqual(rotations, 0, 0.1) {
+		cm.logger.CDebug(ctx, "GoTo distance nearly zero, not moving")
+		return nil
+	}
+	return cm.GoFor(ctx, rpm, rotations, extra)
+}
+
+// GoFor instructs the motor to go in a specific direction for a specific amount of
+// revolutions at a given speed in revolutions per minute. Both the RPM and the revolutions
+// can be assigned negative values to move in a backwards direction. Note: if both are
+// negative the motor will spin in the forward direction.
+// If revolutions is 0, this will run the motor at rpm indefinitely
+// If revolutions != 0, this will block until the number of revolutions has been completed or another operation comes in.
+func (cm *controlledMotor) GoFor(ctx context.Context, rpm, revolutions float64, extra map[string]interface{}) error {
+	cm.opMgr.CancelRunning(ctx)
+	ctx, done := cm.opMgr.New(ctx)
+	defer done()
+
+	switch speed := math.Abs(rpm); {
+	case speed < 0.1:
+		cm.logger.CWarn(ctx, "motor speed is nearly 0 rev_per_min")
+		return motor.NewZeroRPMError()
+	case cm.real.maxRPM > 0 && speed > cm.real.maxRPM-0.1:
+		cm.logger.CWarnf(ctx, "motor speed is nearly the max rev_per_min (%f)", cm.real.maxRPM)
+	default:
+	}
+
+	currentTicks, _, err := cm.enc.Position(ctx, encoder.PositionTypeTicks, extra)
+	if err != nil {
+		return err
+	}
+
+	goalPos, _, _ := encodedGoForMath(rpm, revolutions, currentTicks, cm.ticksPerRotation)
+
+	if cm.loop == nil {
+		// create new control loop if control config exists
+		if len(cm.controlLoopConfig.Blocks) != 0 {
```

Comment:
I don't think we need this if statement, but if we do then we also need to include the case where a control loop isn't configured

Reply:
I believe we do need this for the first GoFor call. if the loop doesn't yet exist, configure it and then do all the necessary updating of control blocks to make the loop do what we want for this GoFor. if it's not the first GoFor call, the loop already exists, therefore we can skip straight to updating the loop accordingly

Reply:
I think we don't need the case where a control loop isn't configured because this motor shouldn't be created if there's no loop and no control params. going to slightly change the constructor to make that clear

Reply:
ah yeah sorry wasn't clear, we def need the `if loop == nil` check, but I didn't think we needed `if len(cm.controlLoopConfig.Blocks) != 0` check since we would always have one if a controlled motor gets created.

Reply:
ah I see yes

Reply:
that makes sense I was not looking at the line you left the comment on

Reply:
and I agree. removing it